File: minisearch/stemmer.py
import re
from functools import lru_cache


class SnowballStemmer:
    """
    Snowball stemmer based on rules described on official snowball website
    https://snowballstem.org/algorithms/english/stemmer.html (from 19.07.2025)

    Tested on official dataset with 100% correctness
    input file - https://github.com/snowballstem/snowball-data/blob/c87231db9e26e7fbc524b7000d720fc882a5dc80/english/voc.txt
    output file - https://github.com/snowballstem/snowball-data/blob/c87231db9e26e7fbc524b7000d720fc882a5dc80/english/output.txt
    """

    VOWELS = ("a", "e", "i", "o", "u", "y")
    DOUBLES = ("bb", "dd", "ff", "gg", "mm", "nn", "pp", "rr", "tt")
    LI_ENDINGS = ("c", "d", "e", "g", "h", "k", "m", "n", "r", "t")
    EXCEPTION_WORDS = ("sky", "news", "howe", "atlas", "cosmos", "bias", "andes")

    PRE_STEM_EXCEPTIONS = {
        "skis": "ski",
        "skies": "sky",
        "idly": "idl",
        "gently": "gentl",
        "ugly": "ugli",
        "early": "earli",
        "only": "onli",
        "singly": "singl",
        "sky": "sky",
        "news": "news",
        "howe": "howe",
    }

    R1_BEGININGS_REGEX = r"^(gener|commun|arsen|past|univers|later|emerg|organ)"

    STEP_1A_SUFFIX_REGEX = r"(sses|ied|ies|us|ss|s)$"

    STEP_2_SUFFIX_MAP = {
        "ization": "ize",
        "ational": "ate",
        "fulness": "ful",
        "ousness": "ous",
        "iveness": "ive",
        "tional": "tion",
        "biliti": "ble",
        "lessli": "less",
        "entli": "ent",
        "ation": "ate",
        "alism": "al",
        "aliti": "al",
        "ousli": "ous",
        "iviti": "ive",
        "ogist": "og",
        "fulli": "ful",
        "enci": "ence",
        "anci": "ance",
        "abli": "able",
        "izer": "ize",
        "ator": "ate",
        "alli": "al",
        "bli": "ble",
        "ogi": "og",
        "li": "",
    }

    STEP_2_SUFFIX_REGEX = r"(ization|ational|fulness|ousness|iveness|tional|biliti|lessli|entli|ation|alism|aliti|ousli|iviti|ogist|fulli|enci|anci|abli|izer|ator|alli|bli|ogi|li)$"

    STEP_3_SUFFIX_MAP = {
        "ational": "ate",
        "tional": "tion",
        "alize": "al",
        "icate": "ic",
        "iciti": "ic",
        "ative": "",
        "ical": "ic",
        "ness": "",
        "ful": "",
    }

    STEP_3_SUFFIX_REGEX = r"(ational|tional|alize|icate|iciti|ative|ical|ness|ful)$"

    STEP_4_SUFFIX_REGEX = r"(ement|ance|ence|able|ible|ment|ant|ent|ism|ate|iti|ous|ive|ize|ion|al|er|ic)$"

    def __init__(self):
        self._r1 = float("inf")
        self._r2 = float("inf")

    @lru_cache(maxsize=1024)
    def stem(self, word):
        """
        Stem the word if it has more than two characters,
        otherwise return it as is.
        """

        if len(word) <= 2 or word in self.__class__.EXCEPTION_WORDS:
            return word

        self._r1, self._r2 = len(word), len(word)

        word = self.remove_initial_apostrophe(word)
        if word in self.__class__.PRE_STEM_EXCEPTIONS:
            return self.__class__.PRE_STEM_EXCEPTIONS[word]

        word = self.set_ys(word)
        self.find_r1r2(word)

        word = self.step_0(word)
        word = self.step_1a(word)
        word = self.step_1b(word)
        word = self.step_1c(word)
        word = self.step_2(word)
        word = self.step_3(word)
        word = self.step_4(word)
        word = self.step_5(word)
        word = word.replace("Y", "y")

        return word

    def find_r1r2(self, word):
        length = len(word)
        self._r1, self._r2 = length, length

        if prefix := re.findall(self.__class__.R1_BEGININGS_REGEX, word):
            self._r1 = len(prefix[0])

            for index, match in enumerate(
                re.finditer("[aeiouy][^aeiouy]", word[self._r1 :])
            ):
                self._r2 = self._r1 + match.end()
                break

        else:
            for index, match in enumerate(re.finditer("[aeiouy][^aeiouy]", word)):
                if index == 0:
                    if match.end() < length:
                        self._r1 = match.end()
                if index == 1:
                    if match.end() < length:
                        self._r2 = match.end()
                    break

    def remove_initial_apostrophe(self, word):
        if word[0] == "'":
            word = word[1:]

        return word

    def set_ys(self, word):

        if word[0] == "y":
            word = "Y" + word[1:]

        for match in re.finditer("[aeiou]y", word):
            y_index = match.end() - 1
            char_list = [x for x in word]
            char_list[y_index] = "Y"
            word = "".join(char_list)

        return word

    def ends_with_short_syllabe(self, word):
        if word == "past":
            return True
        elif len(word) > 2:
            ending = word[len(word) - 3 :]
            if re.match("[^aeiouy][aeiouy][^aeiouwxY]", ending):
                return True
        else:
            if re.match("[aeiouy][^aeiouy]", word):
                return True

        return False

    def is_short(self, word):
        length = len(word)

        if self._r1 >= length:
            return self.ends_with_short_syllabe(word)

        return False

    def step_0(self, word):

        if word.endswith("'s'"):
            return word[:-3]
        elif word.endswith("'s"):
            return word[:-2]
        elif word.endswith("'"):
            return word[:-1]
        else:
            return word

    def step_1a(self, word):
        if suffix := re.findall(self.__class__.STEP_1A_SUFFIX_REGEX, word):
            suffix = suffix[0]

            if suffix == "sses":
                return word[:-2]
            elif suffix in ("ied", "ies"):
                word = word[:-3]

                if len(word) > 1:
                    word += "i"
                else:
                    word += "ie"
                return word
            elif suffix in ("us", "ss"):
                return word
            elif suffix == "s":
                if re.search(r"[aeiouy]+", word[:-2]):
                    return word[:-1]

        return word

    def step_1b(self, word):

        if suffix := re.findall(r"(eedly|eed)$", word):
            suffix = suffix[0]

            if len(word) - len(suffix) >= self._r1 and not any(
                (word.startswith(s) for s in ("proc", "exc", "succ"))
            ):
                word = word[: -len(suffix)] + "ee"

            return word

        elif suffix := re.findall(r"(ingly|edly|ing|ed)$", word):
            suffix = suffix[0]

            # special case for 'ing'
            if suffix == "ing":
                if re.match("^[^aeiouy]y$", word[:-3]):
                    return word[:-4] + "ie"
                elif word[:-3] in ("inn", "out", "cann", "herr", "earr", "even"):
                    return word

            if re.search(r"[aeiouy]+", word[: -len(suffix)]):
                # delete suffix
                word = word[: -len(suffix)]

                if word[-2:] in ("at", "bl", "iz"):
                    return word + "e"
                elif word[-2:] in self.__class__.DOUBLES and not (
                    len(word) == 3 and word[0] in ("a", "e", "o")
                ):
                    return word[:-1]
                elif self.is_short(word):
                    return word + "e"

        return word

    def step_1c(self, word):
        if len(word) > 2 and word[-1] in "yY" and word[-2] not in self.__class__.VOWELS:
            return word[:-1] + "i"

        return word

    def step_2(self, word):

        if suffix := re.findall(self.__class__.STEP_2_SUFFIX_REGEX, word):
            suffix, repl = suffix[0], self.__class__.STEP_2_SUFFIX_MAP[suffix[0]]

            if not (len(word) - len(suffix) >= self._r1):
                return word

            if suffix == "ogi" and word[-4] == "l":
                return word[: -len(suffix)] + repl
            elif suffix == "li" and word[-3] in self.__class__.LI_ENDINGS:
                return word[: -len(suffix)]
            elif suffix not in ["ogi", "li"]:
                return word[: -len(suffix)] + repl

        return word

    def step_3(self, word):

        if suffix := re.findall(self.__class__.STEP_3_SUFFIX_REGEX, word):
            suffix, repl = suffix[0], self.__class__.STEP_3_SUFFIX_MAP[suffix[0]]

            if word.endswith(suffix):

                if not (len(word) - len(suffix) >= self._r1):
                    return word

                if suffix == "ative" and len(word) - len(suffix) >= self._r2:
                    return word[: -len(suffix)] + repl
                elif suffix != "ative":
                    return word[: -len(suffix)] + repl

        return word

    def step_4(self, word):
        if suffix := re.findall(self.__class__.STEP_4_SUFFIX_REGEX, word):
            suffix = suffix[0]

            # r2 in suffix
            if not (len(word) - len(suffix) >= self._r2):
                return word

            if suffix == "ion" and word[-4] in "st":
                return word[: -len(suffix)]
            elif suffix != "ion":
                return word[: -len(suffix)]

        return word

    def step_5(self, word):

        if word.endswith("e"):
            if len(word) - 1 >= self._r2:
                return word[:-1]
            elif len(word) - 1 >= self._r1 and not self.ends_with_short_syllabe(
                word[:-1]
            ):
                return word[:-1]

        elif word.endswith("l") and len(word) - 1 >= self._r2 and word.endswith("ll"):
            return word[:-1]

        return word

File: minisearch/src/analysis/stemmer.rs
use std::collections::HashMap;

static VOWELS: [char; 6] = ['a', 'e', 'i', 'o', 'u', 'y'];
static DOUBLES: [&str; 9] = ["bb", "dd", "ff", "gg", "mm", "nn", "pp", "rr", "tt"];
static LI_ENDINGS: [char; 10] = ['c', 'd', 'e', 'g', 'h', 'k', 'm', 'n', 'r', 't'];
static EXCEPTION_WORDS: [&str; 7] = ["sky", "news", "howe", "atlas", "cosmos", "bias", "andes"];
static R1_BEGININGS: [&str; 8] = [
    "gener", "commun", "arsen", "past", "univers", "later", "emerg", "organ",
];

static STEP_1A_SUFFIXES: [&str; 6] = ["sses", "ied", "ies", "us", "ss", "s"];
static STEP_1B_SUFFIXES_1: [&str; 2] = ["eedly", "eed"];
static STEP_1B_SUFFIXES_2: [&str; 4] = ["ingly", "edly", "ing", "ed"];
static STEP_2_SUFFIXES: [&str; 25] = [
    "ization", "ational", "fulness", "ousness", "iveness", "tional", "biliti", "lessli", "entli",
    "ation", "alism", "aliti", "ousli", "iviti", "ogist", "fulli", "enci", "anci", "abli", "izer",
    "ator", "alli", "bli", "ogi", "li",
];

static STEP_3_SUFFIXES: [&str; 9] = [
    "ational", "tional", "alize", "icate", "iciti", "ative", "ical", "ness", "ful",
];
static STEP_4_SUFFIXES: [&str; 18] = [
    "ement", "ance", "ence", "able", "ible", "ment", "ant", "ent", "ism", "ate", "iti", "ous",
    "ive", "ize", "ion", "al", "er", "ic",
];

pub struct SnowballStemmer {
    r1: usize,
    r2: usize,
    pre_stem_exceptions: HashMap<String, String>,
    step_2_suffix_map: HashMap<String, String>,
    step_3_suffix_map: HashMap<String, String>,
}

impl SnowballStemmer {
    pub fn new() -> Self {
        let pre_stem_exceptions: HashMap<String, String> = HashMap::from([
            (String::from("skis"), String::from("ski")),
            (String::from("skies"), String::from("sky")),
            (String::from("idly"), String::from("idl")),
            (String::from("gently"), String::from("gentl")),
            (String::from("ugly"), String::from("ugli")),
            (String::from("early"), String::from("earli")),
            (String::from("only"), String::from("onli")),
            (String::from("singly"), String::from("singl")),
            (String::from("sky"), String::from("sky")),
            (String::from("news"), String::from("news")),
            (String::from("howe"), String::from("howe")),
        ]);

        let step_2_suffix_map: HashMap<String, String> = HashMap::from([
            (String::from("ization"), String::from("ize")),
            (String::from("ational"), String::from("ate")),
            (String::from("fulness"), String::from("ful")),
            (String::from("ousness"), String::from("ous")),
            (String::from("iveness"), String::from("ive")),
            (String::from("tional"), String::from("tion")),
            (String::from("biliti"), String::from("ble")),
            (String::from("lessli"), String::from("less")),
            (String::from("entli"), String::from("ent")),
            (String::from("ation"), String::from("ate")),
            (String::from("alism"), String::from("al")),
            (String::from("aliti"), String::from("al")),
            (String::from("ousli"), String::from("ous")),
            (String::from("iviti"), String::from("ive")),
            (String::from("ogist"), String::from("og")),
            (String::from("fulli"), String::from("ful")),
            (String::from("enci"), String::from("ence")),
            (String::from("anci"), String::from("ance")),
            (String::from("abli"), String::from("able")),
            (String::from("izer"), String::from("ize")),
            (String::from("ator"), String::from("ate")),
            (String::from("alli"), String::from("al")),
            (String::from("bli"), String::from("ble")),
            (String::from("ogi"), String::from("og")),
            (String::from("li"), String::from("")),
        ]);

        let step_3_suffix_map: HashMap<String, String> = HashMap::from([
            (String::from("ational"), String::from("ate")),
            (String::from("tional"), String::from("tion")),
            (String::from("alize"), String::from("al")),
            (String::from("icate"), String::from("ic")),
            (String::from("iciti"), String::from("ic")),
            (String::from("ative"), String::from("")),
            (String::from("ical"), String::from("ic")),
            (String::from("ness"), String::from("")),
            (String::from("ful"), String::from("")),
        ]);

        SnowballStemmer {
            r1: 0,
            r2: 0,
            pre_stem_exceptions: pre_stem_exceptions,
            step_2_suffix_map: step_2_suffix_map,
            step_3_suffix_map: step_3_suffix_map,
        }
    }

    pub fn stem(&mut self, mut word: String) -> String {
        if word.len() <= 2 || EXCEPTION_WORDS.contains(&word.as_str()) || !word.is_ascii() {
            return word;
        }

        self.remove_initial_apostrophe(&mut word);
        match self.pre_stem_exceptions.get(&word) {
            Some(word) => return word.to_string(),
            None => (),
        }

        self.set_ys(&mut word);
        self.find_r1r2(&mut word);

        self.step_0(&mut word);
        self.step_1a(&mut word);
        self.step_1b(&mut word);
        self.step_1c(&mut word);
        self.step_2(&mut word);
        self.step_3(&mut word);
        self.step_4(&mut word);
        self.step_5(&mut word);
        word.replace("Y", "y")
    }
}

// impl not exposed to python
impl SnowballStemmer {
    fn ends_with_short_syllabe(&self, word: &str) -> bool {
        if word == "past" {
            return true;
        } else if word.len() > 2
            && !VOWELS.contains(&word.chars().nth(word.len() - 3).unwrap())
            && VOWELS.contains(&word.chars().nth(word.len() - 2).unwrap())
            && !['a', 'e', 'i', 'o', 'u', 'w', 'x', 'Y']
                .contains(&word.chars().nth(word.len() - 1).unwrap())
        {
            return true;
        } else if word.len() == 2
            && VOWELS.contains(&word.chars().nth(0).unwrap())
            && !VOWELS.contains(&word.chars().nth(1).unwrap())
        {
            return true;
        }

        return false;
    }

    fn is_short(&self, word: &String) -> bool {
        if self.r1 >= word.len() {
            return self.ends_with_short_syllabe(&word);
        }

        return false;
    }

    fn remove_initial_apostrophe(&self, word: &mut String) {
        if word.starts_with("'") {
            let _ = &word.remove(0);
        }
    }

    fn set_ys(&self, word: &mut String) {
        if word.starts_with("y") {
            word.replace_range(0..1, "Y");
        }

        let mut is_vowel = false;
        let chars: Vec<char> = word.chars().collect();

        for (i, c) in chars.into_iter().enumerate() {
            if is_vowel && c == 'y' {
                word.replace_range(i..i + 1, "Y");
            }

            if VOWELS.contains(&c) {
                is_vowel = true;
            } else {
                is_vowel = false;
            }
        }
    }

    fn find_r1r2(&mut self, word: &mut String) {
        self.r1 = word.len();
        self.r2 = word.len();

        for prefix in R1_BEGININGS.iter() {
            if !word.starts_with(prefix) {
                continue;
            }

            self.r1 = prefix.len();

            let chars: &Vec<char> = &word[self.r1..].chars().collect();
            let mut is_vowel = false;

            for (i, c) in chars.into_iter().enumerate() {
                if VOWELS.contains(&c) {
                    is_vowel = true;
                } else {
                    if is_vowel {
                        self.r2 = self.r1 + i + 1;
                        break;
                    }

                    is_vowel = false;
                }
            }

            return;
        }

        let chars: Vec<char> = word.chars().collect();
        let mut is_vowel = false;
        let mut matches = 0;

        for (i, c) in chars.into_iter().enumerate() {
            if VOWELS.contains(&c) {
                is_vowel = true;
            } else {
                if is_vowel && matches == 0 {
                    matches += 1;
                    self.r1 = i + 1;
                } else if is_vowel && matches == 1 {
                    self.r2 = i + 1;
                    break;
                }

                is_vowel = false;
            }
        }
    }

    fn step_0(&self, word: &mut String) {
        if word.ends_with("'s'") {
            word.truncate(word.len() - 3);
        } else if word.ends_with("'s") {
            word.truncate(word.len() - 2);
        } else if word.ends_with("'") {
            word.truncate(word.len() - 1);
        }
    }

    fn step_1a(&self, word: &mut String) {
        for suffix in STEP_1A_SUFFIXES.iter() {
            if !word.ends_with(suffix) {
                continue;
            }

            if *suffix == "sses" {
                word.truncate(word.len() - 2);
            } else if ["ied", "ies"].contains(&suffix) {
                word.truncate(word.len() - 3);

                if word.len() > 1 {
                    word.insert_str(word.len(), "i");
                } else {
                    word.insert_str(word.len(), "ie");
                }
            } else if *suffix == "s" && word.len() > 2 {
                let chars: &Vec<char> = &word[..word.len() - 2].chars().collect();
                for c in chars.into_iter() {
                    if VOWELS.contains(&c) {
                        word.truncate(word.len() - 1);
                        break;
                    }
                }
            }

            break;
        }
    }

    fn step_1b(&self, word: &mut String) {
        for suffix in STEP_1B_SUFFIXES_1.iter() {
            if !word.ends_with(suffix) {
                continue;
            }

            if word.len() - suffix.len() >= self.r1
                && !["proc", "exc", "succ"].iter().any(|p| word.starts_with(p))
            {
                word.replace_range(word.len() - suffix.len().., "ee");
            }

            return;
        }

        for suffix in STEP_1B_SUFFIXES_2.iter() {
            if !word.ends_with(suffix) {
                continue;
            }

            // special case for 'ing'
            if *suffix == "ing" {
                if word.len() == 5
                    && word.chars().nth(word.len() - 4).unwrap() == 'y'
                    && !VOWELS.contains(&word.chars().nth(word.len() - 5).unwrap())
                {
                    word.replace_range(word.len() - 4.., "ie");
                    return;
                } else if ["inn", "out", "cann", "herr", "earr", "even"]
                    .contains(&&word[..word.len() - 3])
                {
                    return;
                }
            }

            if word[..word.len() - suffix.len()]
                .chars()
                .into_iter()
                .any(|c| VOWELS.contains(&c))
            {
                // delete suffix
                word.truncate(word.len() - suffix.len());

                if ["at", "bl", "iz"].iter().any(|s| word.ends_with(s)) {
                    word.insert(word.len(), 'e');
                } else if DOUBLES.iter().any(|s| word.ends_with(s))
                    && !(word.len() == 3 && ['a', 'e', 'o'].contains(&word.chars().nth(0).unwrap()))
                {
                    word.truncate(word.len() - 1);
                } else if self.is_short(&word) {
                    word.insert(word.len(), 'e');
                }
            }

            break;
        }
    }

    fn step_1c(&self, word: &mut String) {
        if word.len() > 2
            && ['y', 'Y'].contains(&word.chars().nth(word.len() - 1).unwrap())
            && !VOWELS.contains(&word.chars().nth(word.len() - 2).unwrap())
        {
            word.replace_range(word.len() - 1.., "i");
        }
    }

    fn step_2(&self, word: &mut String) {
        for suffix in STEP_2_SUFFIXES.iter() {
            if !word.ends_with(suffix) {
                continue;
            }

            if !(word.len() - suffix.len() >= self.r1) {
                return;
            }

            let repl = self.step_2_suffix_map.get(*suffix).unwrap();
            if word.len() >= 4
                && *suffix == "ogi"
                && word.chars().nth(word.len() - 4).unwrap() == 'l'
            {
                word.replace_range(word.len() - suffix.len().., repl);
            } else if word.len() >= 3
                && *suffix == "li"
                && LI_ENDINGS.contains(&word.chars().nth(word.len() - 3).unwrap())
            {
                word.replace_range(word.len() - suffix.len().., repl);
            } else if !["ogi", "li"].contains(suffix) {
                word.replace_range(word.len() - suffix.len().., repl);
            }

            break;
        }
    }

    fn step_3(&self, word: &mut String) {
        for suffix in STEP_3_SUFFIXES.iter() {
            if !word.ends_with(suffix) {
                continue;
            }

            if !(word.len() - suffix.len() >= self.r1) {
                return;
            }

            let repl = self.step_3_suffix_map.get(*suffix).unwrap();
            if *suffix == "ative" && word.len() - suffix.len() >= self.r2 {
                word.replace_range(word.len() - suffix.len().., repl);
            } else if *suffix != "ative" {
                word.replace_range(word.len() - suffix.len().., repl);
            }

            break;
        }
    }

    fn step_4(&self, word: &mut String) {
        for suffix in STEP_4_SUFFIXES.iter() {
            if !word.ends_with(suffix) {
                continue;
            }

            if !(word.len() - suffix.len() >= self.r2) {
                return;
            }

            if word.len() > 3
                && *suffix == "ion"
                && ['s', 't'].contains(&word.chars().nth(word.len() - 4).unwrap())
            {
                word.truncate(word.len() - suffix.len());
            } else if *suffix != "ion" {
                word.truncate(word.len() - suffix.len());
            }

            break;
        }
    }

    fn step_5(&self, word: &mut String) {
        if word.ends_with("e")
            && ((word.len() - 1 >= self.r2)
                || (word.len() - 1 >= self.r1
                    && !self.ends_with_short_syllabe(&word[..word.len() - 1])))
        {
            word.truncate(word.len() - 1);
        } else if word.ends_with("ll") && word.len() - 1 >= self.r2 {
            word.truncate(word.len() - 1);
        }
    }
}
